process_report/invoices/billable_invoice.py

```diff
@@ -0,0 +1,22 @@
+from dataclasses import dataclass
+
+import process_report.invoices.invoice as invoice
+import process_report.util as util
+
+
+@dataclass
+class BillableInvoice(invoice.Invoice):
+    nonbillable_pis: list[str]
+    nonbillable_projects: list[str]
+    old_pi_filepath: str
+
+    def _prepare(self):
+        self.data = util.remove_nonbillables(
+            self.data, self.nonbillable_pis, self.nonbillable_projects
+        )
+        self.data = util.validate_pi_names(self.data)
+
+    def _process(self):
+        old_pi_df = util.load_old_pis(self.old_pi_filepath)
+        self.data, updated_old_pi_df = util.apply_credits_new_pi(self.data, old_pi_df)
+        util.dump_old_pis(self.old_pi_filepath, updated_old_pi_df)
```

Comment:
I am guilty of doing this myself with code I've written but I think dumping these functions in a `utils` module may not be appropriate here. These functions are clearly preparing and processing the data here, you could create methods in this class or the base class if they could be used in other derived classes. `utils.py` is just too vague. 

`utils.py` is okay for functions like `get_timestamp()`, `compare_dates()` etc. 

@larsks @knikolla any opinions on this?

Reply:
We talked about this on the call today, but I figured I would comment explicitly that I agree with @naved001's comment.

Reply:
I will place these functions in the `BillableInvoice` class then. I was on the fence about where to place these functions before submitting the PR, since I thought they might be used by other invoices in the future, and because I thought I should keep the 'purity' of the Invoice classes to only contain the `_process`, `_prepare`, etc.

Reply:
> I will place these functions in the `BillableInvoice` class then. I was on the fence about where to place these functions before submitting the PR, since I thought they might be used by other invoices in the future, and because I thought I should keep the 'purity' of the Invoice classes to only contain the `_process`, `_prepare`, etc.

No need to worry about preserving the purity of classes. You can and should add as many functions as you need to a class.